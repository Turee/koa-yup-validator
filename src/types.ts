import { TypedSchema } from "yup/es/util/types";
import * as yup from "yup";
import { ValidateOptions } from "yup/lib/types";

export type Validator<
  TBody extends TypedSchema,
  TQuery extends TypedSchema,
  TParams extends TypedSchema,
  THead extends TypedSchema
> = {
  body: TBody;
  headers: THead;
  query: TQuery;
  params: TParams;
};

export type ValidatedProperties<
  TBody extends TypedSchema,
  TQuery extends TypedSchema,
  TParams extends TypedSchema,
  THead extends TypedSchema
> = {
  [k in keyof Validator<TBody, TQuery, TParams, THead>]: yup.Asserts<
    Validator<TBody, TQuery, TParams, THead>[k]
  >;
};

export type ValidatedContext<
  TBody extends TypedSchema,
  TQuery extends TypedSchema,
  TParams extends TypedSchema,
  THead extends TypedSchema
> = { request: ValidatedProperties<TBody, TQuery, TParams, THead> };

export type ValidationErrors = Record<
  keyof Partial<Validator<any, any, any, any>>,
  yup.ValidationError
>;

export class RequestValidationError extends Error {
  constructor(message: string, validationErrors: ValidationErrors) {
    super(message);
    this.validationErrors = validationErrors;
  }

  validationErrors: ValidationErrors;
}

export interface ValidationOptions {
  partial?: boolean;
  yup?: ValidateOptions;
}

export type ValidatorOptions = {
  [k in keyof Partial<Validator<any, any, any, any>>]: ValidationOptions;
};
