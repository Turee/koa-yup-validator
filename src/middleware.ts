import * as yup from "yup";
import Koa, { DefaultState } from "koa";
import {
  RequestValidationError,
  ValidatedContext,
  Validator,
  ValidatorOptions,
} from "./types";
import { validate } from "./validate";
import { ValidationError } from "yup";

const defaultOptions: ValidatorOptions = {
  headers: {
    yup: {
      stripUnknown: false,
    },
  },
};

export function createValidationMiddleware<
  TBody extends yup.SchemaOf<any> = yup.AnyObjectSchema,
  TQuery extends yup.SchemaOf<any> = yup.AnyObjectSchema,
  TParams extends yup.SchemaOf<any> = yup.AnyObjectSchema,
  THead extends yup.SchemaOf<any> = yup.AnyObjectSchema
>(
  validators: Partial<Validator<TBody, TQuery, TParams, THead>>,
  options: ValidatorOptions = defaultOptions
): Koa.Middleware<
  DefaultState,
  ValidatedContext<TBody, TQuery, TParams, THead>
> {
  return async (ctx: Koa.Context, next: Koa.Next) => {
    const request = ctx.request as any;
    const validationErrors: any = {};
    for (const k of Object.keys(validators)) {
      try {
        request[k] = await validate(
          (validators as any)[k],
          (options as any)[k],
          request[k]
        );
      } catch (e) {
        if (e instanceof ValidationError) {
          validationErrors[k] = e;
        } else {
          throw e;
        }
      }
    }
    if (Object.keys(validationErrors).length > 0) {
      throw new RequestValidationError(
        "Request validation failed",
        validationErrors
      );
    }
    return next();
  };
}
