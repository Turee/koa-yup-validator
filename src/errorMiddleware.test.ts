import * as yup from "yup";
import {
  createErrorMiddleware,
  formatRequestValidationError,
} from "./errorMiddleware";
import { RequestValidationError } from "./types";
import Koa from "koa";
import Router from "koa-router";
import { createValidationMiddleware } from "./middleware";
import bodyParser from "koa-bodyparser";
import supertest from "supertest";

const getError = (schema: yup.AnySchema, value: any) => {
  try {
    schema.validateSync(value, { abortEarly: false });
  } catch (e) {
    return new RequestValidationError("validation failed", {
      body: e,
    } as any);
  }
  throw new Error("error expected");
};

describe("formatValidationError", () => {
  const simpleSchema = yup.object().shape({
    key: yup.string().required(),
  });

  const nestedSchema = yup.object().shape({
    stringKey: yup.string().required().url(),
    numberKey: yup.number().required().min(5),
    missingKey: yup.string().required(),
    nestedObject: simpleSchema,
  });

  describe.each([
    ["array schema", yup.array().of(yup.number()), ["asdf"]],
    ["number schema", yup.number(), "asdf"],
    ["nested schema", nestedSchema, { stringKey: 5, numberKey: 1 }],
    [
      "array schema with nested values",
      yup.array().of(nestedSchema),
      [{ stringKey: 5, numberKey: 1 }],
    ],
    [
      "containing array schema",
      yup
        .object()
        .shape({ arr: yup.array().of(nestedSchema).required().min(2) }),
      { arr: [{}] },
    ],
  ])("when given %s", (_, schema, value) => {
    const result = formatRequestValidationError(getError(schema, value));

    it("formats error correctly", () => {
      expect(result).toMatchSnapshot();
    });
  });
});

describe("errorMiddleware", () => {
  describe("when using the error middleware", () => {
    const app = new Koa();

    const schema = yup.object().shape({
      name: yup.string().required(),
      toppings: yup.array().of(yup.string().required()).required(),
    });

    const router = new Router();
    router.post(
      "/hello",
      createValidationMiddleware({
        body: schema,
      }),
      ({ request, response }) => {
        response.body = { count: request.body.ingredients };
      }
    );
    app.use(bodyParser());
    app.use(createErrorMiddleware());
    app.use(router.routes());

    const target = supertest(app.callback());

    describe("when called with invalid request", () => {
      const responsePromise = target.post("/hello").send({});
      it("returns 400", async () => {
        const response = await responsePromise;
        expect(response.status).toEqual(400);
      });
      it("returns formatted error", async () => {
        const response = await responsePromise;
        expect(response.body).toMatchSnapshot();
      });
    });
  });
});
