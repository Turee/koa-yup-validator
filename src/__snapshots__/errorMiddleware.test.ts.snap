// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errorMiddleware when using the error middleware when called with invalid request returns formatted error 1`] = `
Object {
  "type": "ValidationError",
  "validationErrors": Object {
    "body": Array [
      Object {
        "message": "name is a required field",
        "path": "name",
        "type": "required",
      },
      Object {
        "message": "toppings is a required field",
        "path": "toppings",
        "type": "required",
      },
    ],
  },
}
`;

exports[`formatValidationError when given array schema formats error correctly 1`] = `
Object {
  "type": "ValidationError",
  "validationErrors": Object {
    "body": Array [
      Object {
        "message": "[0] must be a \`number\` type, but the final value was: \`NaN\` (cast from the value \`\\"asdf\\"\`).",
        "path": "[0]",
        "type": "typeError",
      },
    ],
  },
}
`;

exports[`formatValidationError when given array schema with nested values formats error correctly 1`] = `
Object {
  "type": "ValidationError",
  "validationErrors": Object {
    "body": Array [
      Object {
        "message": "[0].stringKey must be a valid URL",
        "path": "[0].stringKey",
        "type": "url",
      },
      Object {
        "message": "[0].numberKey must be greater than or equal to 5",
        "path": "[0].numberKey",
        "type": "min",
      },
      Object {
        "message": "[0].missingKey is a required field",
        "path": "[0].missingKey",
        "type": "required",
      },
      Object {
        "message": "[0].nestedObject.key is a required field",
        "path": "[0].nestedObject.key",
        "type": "required",
      },
    ],
  },
}
`;

exports[`formatValidationError when given containing array schema formats error correctly 1`] = `
Object {
  "type": "ValidationError",
  "validationErrors": Object {
    "body": Array [
      Object {
        "message": "arr[0].stringKey is a required field",
        "path": "arr[0].stringKey",
        "type": "required",
      },
      Object {
        "message": "arr[0].numberKey is a required field",
        "path": "arr[0].numberKey",
        "type": "required",
      },
      Object {
        "message": "arr[0].missingKey is a required field",
        "path": "arr[0].missingKey",
        "type": "required",
      },
      Object {
        "message": "arr[0].nestedObject.key is a required field",
        "path": "arr[0].nestedObject.key",
        "type": "required",
      },
      Object {
        "message": "arr field must have at least 2 items",
        "path": "arr",
        "type": "min",
      },
    ],
  },
}
`;

exports[`formatValidationError when given nested schema formats error correctly 1`] = `
Object {
  "type": "ValidationError",
  "validationErrors": Object {
    "body": Array [
      Object {
        "message": "stringKey must be a valid URL",
        "path": "stringKey",
        "type": "url",
      },
      Object {
        "message": "numberKey must be greater than or equal to 5",
        "path": "numberKey",
        "type": "min",
      },
      Object {
        "message": "missingKey is a required field",
        "path": "missingKey",
        "type": "required",
      },
      Object {
        "message": "nestedObject.key is a required field",
        "path": "nestedObject.key",
        "type": "required",
      },
    ],
  },
}
`;

exports[`formatValidationError when given number schema formats error correctly 1`] = `
Object {
  "type": "ValidationError",
  "validationErrors": Object {
    "body": Array [
      Object {
        "message": "this must be a \`number\` type, but the final value was: \`NaN\` (cast from the value \`\\"asdf\\"\`).",
        "path": "",
        "type": "typeError",
      },
    ],
  },
}
`;
